### ISSUE TITLE
Version 1.9 reboots immediately after booting because of millis() overflow protection.

### ISSUE THREAD

**Issue by Nonstle** [roblatour/ESP32RemoteForVictron#4]
**Describe the bug**
Version 1.9 reboots immediately after booting and gets stuck in a boot loop. Not quite sure how it works for you but for me ESP.restart() always gets called the first time "KeepMQTTAlive" is run from "loop".

https://github.com/roblatour/ESP32RemoteForVictron/blob/463db59eca42d1ba0afad5e2c10ff8acaf9799c6/ESP32RemoteForVictron/ESP32RemoteForVictron.ino#L1166-L1168

In my eyes this makes sense because the value "nextTimeCheck" is initialized 0 and only ever appointed a new value in the function "RefreshTimeOnceADay".  and in "loop" "KeepMQTTAlive" is called before "RefreshTimeOnceADay".

And then this statement is true because of course time has already passed since booting the chip:

https://github.com/roblatour/ESP32RemoteForVictron/blob/463db59eca42d1ba0afad5e2c10ff8acaf9799c6/ESP32RemoteForVictron/ESP32RemoteForVictron.ino#L1167

"RefreshTimeOnceADay" is called when a wifi connection is made but that must be to late, so wifi never gets the opportunity to complete.

**Possible solution**
I haven't looked through the rest of the code very closely but if you're using "millis" the correct way and only with unsigned integers there is no need to reboot every 49 days. Example:
```
if (millis - lastTimeActionPerformed >= interval)
```
This will work even if there was an overflow and lastTimeActionPerformed is bigger then millis at that point, because it takes advantage of the fact the minus calculation will overflow as well, giving us still the expected value of how much time has passed.

**Other info**
I'm running it on a LILYGO T-Display-S3 AMOLED module. 
Awesome project by the way, thanks for sharing.

**Pull Request**
I'd like to look through the rest of the code and possibly fix any checks involving millis and make a pull request without the need to reboot every 49 days so you can use it. But I thought I'd ask if you're interested before putting in the effort.

**Comment by roblatour** [comment]
Thanks for pointing this out, I just looked at my code and I had GENERAL_SETTINGS_SEND_PERIODICAL_KEEP_ALIVE_REQUESTS   set to false, and why I missed it in testing.  Let me take a closer look and see if I can find a solution.  Sorry about the inconvenience.

**Comment by Nonstle** [comment]
Ah yeah that will do it, no worries man and no rush. I removed the reboots for now and it's working nicely.

**Comment by roblatour** [comment]
ok - I just updated the code to version 1.9.1 with a fix.  

I wasn't sure what you meant by "if you're using "millis" the correct way" - so maybe I am and maybe I am not :-)

Regardless, it was a simple change and it tested just fine (with GENERAL_SETTINGS_SEND_PERIODICAL_KEEP_ALIVE_REQUESTS set to true, and also when set to false).

If your currently running with the code, you will not need to adjust your settings - just replace the old .ino file with the new one and rebuild and upload to your device.

Would appreciate your feedback if this fixes it for you, and if yes I'd invite you to please close this issue. 

Again, with thanks.

**Comment by Nonstle** [comment]
>ok - I just updated the code to version 1.9.1 with a fix.

Awesome, just tested it and no more boot loop!

>I wasn't sure what you meant by "if you're using "millis" the correct way" - so maybe I am and maybe I am not :-)

Haha, I don't know the history of the project of course but I assume there was a reason you implemented the overflow reboot routine in 1.9 so that's why I brought it up.

I took a look and for example this is the first time i found "millis" being used:

https://github.com/roblatour/ESP32RemoteForVictron/blob/0d4e8a17005ce9a536d3c07e034d41adc36f72b0/ESP32RemoteForVictron/ESP32RemoteForVictron.ino#L237

And this indeed leaves you vulnerable for an overflow, if "millis" itself or "millis" plus the 60000 is more then an unsigned long can hold this calculation will overflow and the if statement will never be true anymore for at least the 49 days.

This is because you're using it with exact numbers, as in "I want something to happen when we're past timestamp x". 

You'll be safe from overflows if you use it relatively, as in "I want something to happen if more then x amount of time has passed"

My little example expanded, this is safe from overflows.

```
unsigned long lastTimeExecuted = 0;

loop() {
    if (millis() - lastTimeExecuted >= 1000) {
        lastTimeExecuted = millis();

        //do something that needs to be done every 1000 miliseconds
    }
}
```

Hope this helps and like I said, I'm happy to take a swing at it and make a pull request, just let me know.

Thank you for your quick response. I'll close this issue.

Fixed in: 0d4e8a17005ce9a536d3c07e034d41adc36f72b0

**Comment by roblatour** [comment]
@Nonstle - thanks for letting me know - glad its working for you.  

As for the following example:

unsigned long lastTimeExecuted = 0;

loop() {
    if (millis() - lastTimeExecuted >= 1000) {
        lastTimeExecuted = millis();

        //do something that needs to be done every 1000 miliseconds
    }
}

I'm not fully sure how that would solve the problem, eventually (after 49 days) millis() would role over and when that happens it would take another 49 days until (millis() - lastTimeExecuted >= 1000) was true, if it happened at all.  Perhaps we can continue the discussion outside of git, you're welcome to email me at info@rlatour.com if you like.
